service/src/main/java/io/camunda/service/UserTaskServices.java

```diff
@@ -40,11 +58,176 @@ public UserTaskServices withAuthentication(final Authentication authentication)
 
   @Override
   public SearchQueryResult<UserTaskEntity> search(final UserTaskQuery query) {
+    if (query.filter().variableFilters() != null && !query.filter().variableFilters().isEmpty()) {
+      final Map<Long, Set<Long>> scopeKeyMapByProcessInstanceKey = new HashMap<>();
+      final VariableServices variableServices =
+          new VariableServices(searchClient, transformers, authentication);
+      boolean firstIteration = true;
+
+      for (final VariableValueFilter variableValueFilter : query.filter().variableFilters()) {
+        processVariableFilter(
+            variableValueFilter, variableServices, scopeKeyMapByProcessInstanceKey, firstIteration);
+        firstIteration = false;
+      }
+
+      // Collect all unique processInstanceIds and scopeKeys
+      final List<Long> allProcessInstanceIds =
+          new ArrayList<>(scopeKeyMapByProcessInstanceKey.keySet());
+      final List<Long> allScopeKeys =
+          scopeKeyMapByProcessInstanceKey.values().stream()
+              .flatMap(set -> Optional.ofNullable(set).orElse(Collections.emptySet()).stream())
+              .filter(Objects::nonNull)
+              .collect(Collectors.toList());
+
+      if (allProcessInstanceIds.isEmpty() && allScopeKeys.isEmpty()) {
+        return new SearchQueryResult<>(0, Collections.emptyList(), null);
+      }
+
+      final UserTaskFilter mergedFilter =
+          getUserTaskFilter(query, allProcessInstanceIds, allScopeKeys);
+
+      // Create a new UserTaskQuery with the merged filter
+      final UserTaskQuery modifiedQuery =
+          UserTaskQuery.of(b -> b.filter(mergedFilter).sort(query.sort()).page(query.page()));
+
+      return executor.search(modifiedQuery, UserTaskEntity.class);
+    }
     return executor.search(query, UserTaskEntity.class);
```

Comment:
:x: Please adjust the code in a way that the statement `executor.search(...)` is only there once.